### PR TITLE
Demote UTC to I/O format and general overhaul

### LINKED ISSUE #50: Error and strange behaviour around the previous day of leap-second's day
Hi, I am a new user of AstroTime.jl, and I found this package very useful.  I encountered an issue as below.  Note that a leap second was inserted before 2017-1-1 00:00:00 UTC, but this is the case around 2016-12-31 00:00:00 UTC.  The same happens for 2015-6-30.

```
julia> ep = UTCEpoch(2016, 12, 31, 0, 0, 0.0)
2016-12-30T23:59:59.000 UTC

julia> ep = UTCEpoch(2016, 12, 31, 0, 0, 0.1)
Error showing value of type Epoch{CoordinatedUniversalTime,Float64}:
ERROR: ArgumentError: Seconds are out of range
Stacktrace:
(followed by a long message)

julia> ep = UTCEpoch(2016, 12, 31, 0, 1, 0.0)
2016-12-31T00:00:60.000 UTC
```

Thanks, 
Toshi

### LINKED ISSUE #53: now() returns local time as UTCEpoch instead of UTC
Hi,
I'm too dumb to use git or I would have created an actual pull request (if that's the proper term).

AstroTime.now() returns the computer's local time as it were UTC.
In the file Epochs.jl, line 352ish, I've changed
- now() = UTCEpoch(Dates.now())
into
+ now() = UTCEpoch(Dates.now(Dates.UTC))
And that change makes AstroTime.now() return the UTC time independently of your local time.

### PR DESCRIPTION
**Pull request by helgee**
**Boy, that escalated quickly!**

I realize that this is a lot to review so I will leave comments inline to highlight the most important changes. Documentation updates will follow in a separate PR.

Fixes #50
Fixes #53

### CODE REVIEW

File: src/AstroDates.jl [hunk 1]
```diff
@@ -216,113 +175,235 @@ const UNIX_EPOCH = Date(1970, 1, 1)
...
-Dates.Time(t::Time) = Dates.Time(hour(t), minute(t), second(t), millisecond(t))
+function Base.isapprox(a::Time, b::Time; kwargs...)
+    return a.hour == b.hour &&
+           a.minute == b.minute &&
+           isapprox(a.fraction + a.second, b.fraction + b.second; kwargs...)
+end
 
-const H00 = Time(0, 0, 0.0)
-const H12 = Time(12, 0, 0.0)
+function subsecond(fraction, n, r)
```
Comment (helgee) [inline src/AstroDates.jl:220]: This stumped me for a while. See also the method below. As a result, we now get sensible values for milli-/micro-/nanoseconds as if they were integral values and hiding the inherent floating point inaccuracies.

```julia
julia> subsecond(0.1839999999999975, 9)
0

julia> subsecond(0.1839999999999975, 3)
184

julia> subsecond(0.123456789, 9)
789

julia> subsecond(0.123456789, 3)
123
```

**Review by giordano — APPROVED** [review]
I looked more the implementation than the merit of the changes :slightly_smiling_face:

File: src/Epochs/io.jl [hunk 2]
```diff
@@ -1,25 +1,33 @@
...
 
 function Dates.format(io, d::Dates.DatePart{'D'}, ep)
     print(io, dayofyear(ep))
 end
 
-function Dates.format(io, d::Dates.DatePart{'t'}, ep)
-    print(io, timescale(ep))
+abstract type FractionOfSecondToken end
+
+@inline function Dates.tryparsenext(d::Dates.DatePart{'f'}, str, i, len, locale)
```
Comment (helgee) [inline src/Epochs/io.jl:21]: We can now parse timestamps with full `Float64` precision by using the `f` field, e.g. `dateformat"yyyy-mm-ddTHH:MM:SS.ffffffff`. Thereby circumventing https://github.com/JuliaLang/julia/issues/37579

File: src/Epochs/offsets.jl [hunk 3]
```diff
@@ -295,140 +295,51 @@ function getoffset(::BarycentricDynamicalTime, ::BarycentricCoordinateTime, seco
...
 #######
 # UT1 #
 #######
 
 """
-    getoffset(UTC, UT1, second, fraction[, eop])
+    getoffset(TAI, UT1, second, fraction[, eop])
 
-Return the offset between [`UTC`](@ref) and [`UT1`](@ref) for the
+Return the offset between TAITC`](@ref) and [`UT1`](@ref) for the
```
Comment (helgee) [inline src/Epochs/offsets.jl:305]:
```suggestion
Return the offset between [`TAI`](@ref) and [`UT1`](@ref) for the
```

File: src/Epochs/offsets.jl [hunk 4]
```diff
@@ -295,140 +295,51 @@ function getoffset(::BarycentricDynamicalTime, ::BarycentricCoordinateTime, seco
...
 
 ```jldoctest; setup = :(using AstroTime; AstroTime.load_test_eop())
-julia> getoffset(UTC, UT1, 0, 0.0)
-0.3550253556501879
+julia> getoffset(TAI, UT1, 0, 0.0)
+-31.644974644349812
 ```
 """
-@inline function getoffset(::CoordinatedUniversalTime, ::UniversalTime,
+@inline function getoffset(::InternationalAtomicTime, ::UniversalTime,
```
Comment (helgee) [inline src/Epochs/offsets.jl:317]: UT1 is linked to TAI now.

File: src/Epochs/utc.jl [hunk 5]
```diff
@@ -0,0 +1,90 @@
+const LEAP_J2000 = round.(Int, (LeapSeconds.LS_EPOCHS .- value(J2000_TO_MJD)) * 86400)
+const LEAP_TAI = LEAP_J2000 .+ round.(Int, LeapSeconds.LEAP_SECONDS) .- 1
+const LEAP_TAI_SET = Set(LEAP_TAI)
+
+function from_utc(str::AbstractString;
```
Comment (helgee) [inline src/Epochs/utc.jl:5]: This is the headline change. There is no `Epoch{UTC}` anymore but we can still convert to/from UTC via the dedicated `from_utc` and `to_utc`. The discussion about this is in https://github.com/JuliaAstro/AstroTime.jl/issues/50#issuecomment-662049786.

**TL;DR:** The `Epoch` data format is incompatible with a discontinuous timescale like UTC (and only UTC 👎 ). Thus, it makes no sense to rewrite the whole library and add a lot of edge cases for UTC. The best way of dealing with UTC and leap seconds is by keeping them on the outer edge of your code or pretend they do not exist like Google (https://github.com/google/unsmear). *Friends do not let friends use UTC for mission-critical time keeping.*

File: src/Epochs/utc.jl [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+end
+
+function to_utc(::Type{String}, ep, dateformat=Dates.default_format(DateTime))
+    dt = to_utc(DateTime, ep)
+    return Dates.format(dt, dateformat)
+end
+
+to_utc(ep, args...) = to_utc(String, ep, args...)
+
+"""
```
Comment (helgee) [inline src/Epochs/utc.jl:68]: `now` returns actual TAI (from UTC) instead of local time. No idea what will happen during a leap second though 😬

File: test/epochs.jl [hunk 7]
```diff
@@ -157,145 +167,83 @@ end
...
-        @test first(rng) == UTCEpoch(2018, 1, 1)
-        @test last(rng) == UTCEpoch(2018, 2, 1)
-        rng = UTCEpoch(2018, 1, 1):13seconds:UTCEpoch(2018, 1, 1, 0, 1)
+        @test first(rng) == TAIEpoch(2018, 1, 1)
+        @test last(rng) == TAIEpoch(2018, 2, 1)
+        rng = TAIEpoch(2018, 1, 1):13seconds:TAIEpoch(2018, 1, 1, 0, 1)
         @test step(rng) == 13seconds
-        @test last(rng) == UTCEpoch(2018, 1, 1, 0, 0, 52.0)
-    end
-    @testset "Leap Seconds" begin
```
Comment (helgee) [inline test/epochs.jl:212]: All of these become obsolete by removing UTC 😅

File: test/runtests.jl [hunk 8]
```diff
@@ -1,24 +1,24 @@
...
-furnsh(joinpath("data", "naif0012.tls"))
 AstroTime.load_test_eop()
 
 const speed_of_light = 299792458.0 # m/s
 const astronomical_unit = 149597870700.0 # m
 
-@timescale GMT UTC
-AstroTime.Epochs.getoffset(::GMTScale, ::CoordinatedUniversalTime, _, _) = 0.0
-AstroTime.Epochs.getoffset(::CoordinatedUniversalTime, ::GMTScale, _, _) = 0.0
+# TODO: Change me
```
Comment (helgee) [inline test/runtests.jl:11]: These examples are not great anymore and I will try to come up with new ones during the doc update.

File: src/Epochs/offsets.jl (hunk 3 above)
[the remark this answers is not in the document]
Comment (helgee) [reply src/Epochs/offsets.jl]: See below 😉

File: src/AstroDates.jl [hunk 9]
```diff
@@ -1,42 +1,27 @@
...
+using Dates: year, month, day, hour, minute, second
+using Dates: millisecond, microsecond, nanosecond
+using Dates: dayofyear
 
-    return year
-end
+const J2000 = 2.4515445e6
 
-function findyear(calendar, j2000day::Int32)
+function findyear(calendar, j2000day)
```
[the remark this answers is not in the document]
Comment (helgee) [reply src/AstroDates.jl:12]: I am not sure about this. It seems that `Int64(j2000day)` allocates and I saw lots `jl_box_int64` in my profiles. This is ugly but performance is an order of magnitude better: `j2kday = ifelse(j2000day isa Int32, widen(j2000day), j2000day)`

[the remark this answers is not in the document]
Comment (helgee) [reply src/AstroDates.jl:12]: And it is `Int` so changes only with `Sys.WORD_SIZE`.
